index.html

```diff
@@ -410,6 +463,94 @@ <h3 id="installation-sec">
       <section>
         <h3>
+          <code>BeforeInstallPromptEvent</code>
+        </h3>
+        <pre class="idl">
+enum AppBannerPromptOutcome {
+  "accepted",
+  "dismissed"
+};
+
+[Constructor(DOMString typeArg, optional BeforeInstallPromptEventInit eventInit)]
+interface BeforeInstallPromptEvent : Event {
+    readonly attribute Promise&lt;AppBannerPromptOutcome&gt; userChoice;
+    Promise&lt;void&gt; prompt();
+};
+
+dictionary BeforeInstallPromptEventInit : EventInit {
+  AppBannerPromptOutcome userChoice;
+};
+      </pre>
+        <p>
+          The <dfn><code>BeforeInstallPromptEvent</code></dfn> is dispatched
+          prior to activating an <a>automated install prompt</a>, allowing a
+          developer to prevent the default action for an install prompt.
+        </p>
+        <p>
+          Thus, the default action of the <a>BeforeInstallPromptEvent</a> is to
+          present an automated install prompt to the end user. Cancelling the
+          default action (via <code>.preventDefault()</code>) prevents the user
+          agent from presenting the automated install prompt.
+        </p>
+        <p>
+          To create an instance of the <code>BeforeInstallPromptEvent</code>
+          interface, run the <dfn>steps to construct a BeforeInstallPromptEvent
+          event</dfn>:
+        </p>
+        <ol>
+          <li>Let <var>userChoice</var> be null.
+          </li>
+          <li>Let [[\prompCalled]] be false.
+          </li>
+          <li>Let <var>userChoicePromise</var> be a newly created promise.
+          </li>
+          <li>If <code>eventInit</code> was passed:
+            <ol>
+              <li>If <code>eventInit.userChoice</code> was set, then set <var>
+                userChoice</var> to the value of
+                <code>eventInit.userChoice</code>.
+              </li>
+              <li>Resolve <var>userChoicePromise</var> with
+              <var>userChoice</var>.
+              </li>
+            </ol>
+          </li>
+        </ol>
+        <section>
+          <h4>
+            <code>prompt()</code> method
+          </h4>
+          <p>
+            The <dfn><code>prompt()</code> method</dfn>, when called, runs the
+            following steps:
+          </p>
+          <ol>
+            <li>If the event's cancel flag is not set, return a promise
+            rejected with an "InvalidStateError".
+            </li>
+            <li>If [[\]], return a promise rejected with an
+            "InvalidStateError".
+            </li>
+            <li>Let <var>p</var> be a new promise.
+            </li>
+            <li>Run the following steps in parallel:
+            </li>
+            <li>Return <var>p</var>.
+            </li>
+          </ol>
+        </section>
+        <section>
+          <h4>
+            <code>userChoice</code> attribute
+          </h4>
+          <p>
+            The <dfn><code>userChoice</code> attribute</dfn>, when getting,
+            returns a promise.
```

Comment:
Agree. This is painful. 

My proposal would be that BIP has one internal "prompt" promise. That makes the whole design much more sensical. That promise resolves after prompt() succeeds and returns some kind of `UserPromptResult` object as proposed above.